core/src/memory/coalescing_memory_resource.cpp

```diff
@@ -0,0 +1,68 @@
+/**
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "vecmem/memory/coalescing_memory_resource.hpp"
+
+#include <cstddef>
+#include <initializer_list>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+namespace vecmem {
+coalescing_memory_resource::coalescing_memory_resource(
+    std::initializer_list<std::reference_wrapper<memory_resource>> &&upstreams)
+    : m_upstreams(upstreams) {}
+
+void *coalescing_memory_resource::do_allocate(std::size_t size,
+                                              std::size_t align) {
+    /*
+     * Try to allocate with each of the upstream resources.
+     */
+    for (memory_resource &res : m_upstreams) {
+        try {
+            /*
+             * Try to allocate the memory, and store the result with the
+             * allocator reference in the allocation map.
+             */
+            void *ptr = res.allocate(size, align);
+
+            m_allocations.emplace(ptr, res);
+
+            return ptr;
+        } catch (std::bad_alloc &) {
+            /*
+             * If we cannot allocate with this resource, try the next one.
+             */
+            continue;
+        }
+    }
+
+    /*
+     * If all resources fail to allocate, then we do as well.
+     */
+    throw std::bad_alloc();
+}
+
+void coalescing_memory_resource::do_deallocate(void *ptr, std::size_t size,
+                                               std::size_t align) {
+    /*
+     * Fetch the resource used to allocate this pointer.
+     */
+    auto nh = m_allocations.extract(ptr);
+
+    /*
+     * If we know who allocated this memory, forward the deallocation request
+     * to them.
+     */
+    if (nh) {
+        memory_resource &res = nh.mapped();
+
+        res.deallocate(nh.key(), size, align);
+    }
```

Comment:
Should we treat it as an error if an unknown pointer is given to this function for deallocation?

Reply:
I'm not sure if throwing an exception is really a good idea. The standard dictates that the deallocation shouldn't throw exceptions (although this is not enforced through `nothrow`, I think). We might get into some tricky situations if we start throwing exceptions here.

Rather the standard says that trying to deallocate memory that was not allocated with the same (or an equipotent) resource is undefined behaviour, which I think this is as well.